### PR TITLE
Maven 3 upgrade

### PR DESCRIPTION
**Pull request by basil**
I got a little further than #23, although `hpi:run` still isn't working. I could use some help debugging that. I did not bother trying to get `BundlePluginsMojo` to compile, hence this PR being downstream of #246.

### CODE REVIEW

**Comment by basil** [comment]
I have this _mostly_ working, but there is still a problem with `HplMojo#buildLibraries`: it blows up on https://github.com/jenkinsci/maven-hpi-plugin/blob/2f70c5522e3bf1b60c4821e0b5b07da758c2bac3/src/main/java/org/jenkinsci/maven/plugins/hpi/HplMojo.java#L142 because `artifact.getDependencyTrail()` returns a valid trail on the main branch but returns an invalid trail with only one element (which is not expected) with the changes from this PR. That is, this code works on the main branch but fails with the changes from this PR:

https://github.com/jenkinsci/maven-hpi-plugin/blob/2f70c5522e3bf1b60c4821e0b5b07da758c2bac3/src/main/java/org/jenkinsci/maven/plugins/hpi/HplMojo.java#L139-L141

@jglick @Vlatombe I'll keep digging but any help would be appreciated.

File: src/main/java/org/jenkinsci/maven/plugins/hpi/HplMojo.java [hunk 1]
```diff
@@ -137,6 +136,9 @@ private void buildLibraries(List<String> paths) throws IOException {
         for (MavenArtifact artifact : artifacts) {
             if(jenkinsPlugins.contains(artifact.getId()))
                 continue;   // plugin dependencies
+            if (artifact.getDependencyTrail().size() < 2) {
+                throw new IllegalStateException("invalid dependency trail");
+            }
             if(artifact.getDependencyTrail().size() >= 1 && jenkinsPlugins.contains(artifact.getDependencyTrail().get(1)))
```
Comment (Vlatombe) [inline src/main/java/org/jenkinsci/maven/plugins/hpi/HplMojo.java:142]: It's been wrong for 8 years...

```suggestion
            if(artifact.getDependencyTrail().size() > 1 && jenkinsPlugins.contains(artifact.getDependencyTrail().get(1)))
```

File: src/main/java/org/jenkinsci/maven/plugins/hpi/RunMojo.java [hunk 2]
```diff
@@ -783,22 +801,39 @@ public void startJetty() throws MojoExecutionException {
...
+
+            Set<Artifact> artifacts = new LinkedHashSet<>();
+            if (result.getDependencyGraph() != null
+                    && !result.getDependencyGraph().getChildren().isEmpty()) {
+                RepositoryUtils.toArtifacts(
+                        artifacts,
+                        result.getDependencyGraph().getChildren(),
+                        Collections.singletonList(project.getArtifact().getId()),
+                        request.getResolutionFilter());
+            }
```
Comment (Vlatombe) [inline src/main/java/org/jenkinsci/maven/plugins/hpi/RunMojo.java:817]: Did you find what was wrong with the maven 3 way of doing things? Bug on maven side? It looks weird that the dependency trail wouldn't be populated without this.

**Comment by basil** [comment]
I got this working… turns out I had to use `ProjectDependenciesResolver` rather than `DependencyResolver` (go figure). Anyway, integration tests are now passing, and I tested this with `mvn verify` and `mvn hpi:run` in two plugins, and all seems to work, so I am marking this ready for review.

File: src/main/java/org/jenkinsci/maven/plugins/hpi/RunMojo.java (hunk 2 above)
Comment (basil) [reply src/main/java/org/jenkinsci/maven/plugins/hpi/RunMojo.java:819]: >the maven 3 way of doing things

Does not make sense. The comment to `RunMojo#resolveDependencies` says "Performs the equivalent of "@requireDependencyResolution" mojo attribute", and the implementation of `requiresDependencyResolution` in `LifecycleDependencyResolver` uses `ProjectDependenciesResolver`, so my revised code (_not_ the original) _is_ the "Maven 3 way of doing things".

File: src/main/java/org/jenkinsci/maven/plugins/hpi/HplMojo.java (hunk 1 above)
Comment (basil) [reply src/main/java/org/jenkinsci/maven/plugins/hpi/HplMojo.java:142]: I am not touching this pre-existing problem. Instead I added an assertion a line above to make sure the state is as we expect. This assertion passes both before and after this PR.

**Comment by basil** [comment]
@Vlatombe I made some fix-ups to the original changes. Can you please take another look? Thanks!

**Comment by Vlatombe** [comment]
I'm not really able to tell since you force-pushed on your previous changes. Will try on some plugins.

**Review by Vlatombe — APPROVED** [review]
LGTM

**Pull request merged**